staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package shufflesharding
+
+import (
+	"fmt"
+	"math"
+)
+
+// MaxHashBits is the max bit length which can be used from hash value.
+// If we use all bits of hash value, the critical(last) card shuffled by
+// Dealer will be uneven to 2:3 (first half:second half) at most,
+// in order to reduce this unevenness to 32:33, we set MaxHashBits to 60 here.
+const MaxHashBits = 60
+
+// RequiredEntropyBits will figure out the required entropy bits for the given
+// deckSize and handSize, and this value must be less or equal than MaxHashBits
+// if we want to create and use Dealer.
+func RequiredEntropyBits(deckSize, handSize int) int {
+	return int(math.Ceil(math.Log2(float64(deckSize)) * float64(handSize)))
+}
+
+// Dealer contains some necessary parameters and provides some methods for shuffle sharding.
+type Dealer struct {
+	deckSize int
+	handSize int
+
+	// hand and rawCards are pre-allocated to improve performance
```

Comment:
Putting them here makes the Dealer non-thread-safe.  That is, it can not be used concurrently by multiple threads.  That is a problem.  We want a Dealer to be usable by multiple threads concurrently.

Reply:
+1 these can't be here. I don't even think you can use e.g. a sync.Pool to reuse, since you don't know how long a caller will use them for.

Reply:
@MikeSpreitzer @lavalamp Agree on the thread-safe on Dealer, but still have a question, since `hand` will be used by caller **after** shuffle process and `rawCards` is only used **in** shuffle process, could we use sync.Pool for `rawCards` and allocate int slice for `hand` every time?

Reply:
No.  Make it work like it does in the feature branch.  The lack of escape of the pointers should be easy for the compiler to figure out, so it can produce efficient code.  That way we have simple, direct source code and good emitted code.